python/ray/workflow/tests/test_checkpoint.py

```diff
@@ -0,0 +1,164 @@
+          f"recover_duration = {recover_duration_partial}")
+    print(f"[whole] run_duration = {run_duration_whole}, "
+          f"recover_duration = {recover_duration_whole}")
+    assert run_duration_skipped < run_duration_whole
+    assert run_duration_partial < run_duration_whole
+    assert recover_duration_partial < recover_duration_skipped
+    assert recover_duration_whole < recover_duration_skipped
+
+
+@workflow.step
+def identity2(x):
+    if not utils.check_global_mark():
+        import os
+        os.kill(os.getpid(), 9)
+    return x
+
+
+@workflow.step
+def checkpoint_dag2(checkpoint):
+    x = large_input.options(checkpoint=checkpoint).step()
+    y = identity2.options(checkpoint=checkpoint).step(x)
+    return average.step(y)
+
+
+def test_checkpoint_dag_recovery(workflow_start_regular):
+    utils.set_global_mark()
+    # warm up
+    for _ in range(3):
+        outputs = checkpoint_dag2.step(True).run()
+        assert np.isclose(outputs, 8388607.5)
+
+    utils.unset_global_mark()
+
+    start = time.time()
+    with pytest.raises(RaySystemError):
+        checkpoint_dag2.options(checkpoint=False).step(False).run(
+            workflow_id="checkpoint_skip2")
+    run_duration_skipped = time.time() - start
+
+    utils.set_global_mark()
+
+    start = time.time()
+    recovered = ray.get(workflow.resume("checkpoint_skip2"))
+    recover_duration_skipped = time.time() - start
+    assert np.isclose(recovered, 8388607.5)
+
+    utils.unset_global_mark()
+
+    start = time.time()
+    with pytest.raises(RaySystemError):
+        checkpoint_dag2.step(False).run(workflow_id="checkpoint_partial2")
+    run_duration_partial = time.time() - start
+
+    utils.set_global_mark()
+
+    start = time.time()
+    recovered = ray.get(workflow.resume("checkpoint_partial2"))
+    recover_duration_partial = time.time() - start
+    assert np.isclose(recovered, 8388607.5)
+
+    utils.unset_global_mark()
+
+    start = time.time()
+    with pytest.raises(RaySystemError):
+        checkpoint_dag2.step(True).run(workflow_id="checkpoint_whole2")
+    run_duration_whole = time.time() - start
+
+    utils.set_global_mark()
+
+    start = time.time()
+    recovered = ray.get(workflow.resume("checkpoint_whole2"))
+    recover_duration_whole = time.time() - start
+    assert np.isclose(recovered, 8388607.5)
+
+    print(f"[skipped] run_duration = {run_duration_skipped}, "
+          f"recover_duration = {recover_duration_skipped}")
+    print(f"[partial] run_duration = {run_duration_partial}, "
+          f"recover_duration = {recover_duration_partial}")
+    print(f"[whole] run_duration = {run_duration_whole}, "
+          f"recover_duration = {recover_duration_whole}")
+    # We do not compare the running and recovery time here
+    # because due to failures the time is not consistent.
+
+
+def test_checkpoint_dag_validation(workflow_start_regular):
+    # check invalid configuration
+    with pytest.raises(RaySystemError):
```

Comment:
Can we check that the error contains the right message (checking for ValueError is fine)?

Also, could we check that it works if only the output step has checkpoint=True?

Reply:
It seems Ray workflow does not forward internal errors. I am trying to figure out how to capture the error message.